In [1]:
# Import necessary libraries
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils

# Set seed for reproducibility
np.random.seed(1671)

# Network parameters
NB_EPOCH = 20
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10  # number of outputs = number of digits
OPTIMIZER = SGD()  # optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT = 0.2  # how much TRAIN is reserved for VALIDATION

# Load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()
RESHAPED = 784

# Reshape and normalize data
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

# Convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)

# Define model
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
              optimizer=OPTIMIZER,
              metrics=['accuracy'])

# Train the model
history = model.fit(X_train, Y_train,
                    batch_size=BATCH_SIZE, epochs=NB_EPOCH,
                    verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

# Evaluate on test data
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score:", score[0])
print('Test accuracy:', score[1])


Using TensorFlow backend.


11493376/11490434 [==============================] - 0s 0us/step
Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 11s 220us/step - loss: 1.4829 - accuracy: 0.6231 - val_loss: 0.7584 - val_accuracy: 0.8286
Epoch 2/20
48000/48000 [==============================] - 13s 273us/step - loss: 0.6049 - accuracy: 0.8464 - val_loss: 0.4550 - val_accuracy: 0.8852
Epoch 3/20
48000/48000 [==============================] - 13s 272us/step - loss: 0.4398 - accuracy: 0.8801 - val_loss: 0.3710 - val_accuracy: 0.9019
Epoch 4/20
48000/48000 [==============================] - 12s 259us/step - loss: 0.3767 - accuracy: 0.8952 - val_loss: 0.3322 - val_accuracy: 0.9082
Epoch 5/20
48000/48000 [==============================] - 13s 271us/step - loss: 0.3415 - accuracy: 0.9025 - val_loss: 0.3055 - val_accuracy: 0.9147
Epoch 6/20
48000/48000 [==============================] - 14s 285us/step - loss: 0.3175 - accuracy: 0.9086 - val_loss: 0.2880 - val_accuracy

In [2]:
# Experiment 1: Varying Number of Epochs
NB_EPOCH = 10  # Reduce epochs

model.compile(loss='categorical_crossentropy',
              optimizer=OPTIMIZER,
              metrics=['accuracy'])

history = model.fit(X_train, Y_train,
                    batch_size=BATCH_SIZE, epochs=NB_EPOCH,
                    verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score with 10 epochs:", score[0])
print('Test accuracy with 10 epochs:', score[1])


Train on 48000 samples, validate on 12000 samples
Epoch 1/10
48000/48000 [==============================] - 5s 95us/step - loss: 0.1839 - accuracy: 0.9475 - val_loss: 0.1818 - val_accuracy: 0.9502
Epoch 2/10
48000/48000 [==============================] - 4s 86us/step - loss: 0.1791 - accuracy: 0.9486 - val_loss: 0.1771 - val_accuracy: 0.9509
Epoch 3/10
48000/48000 [==============================] - 4s 92us/step - loss: 0.1747 - accuracy: 0.9502 - val_loss: 0.1742 - val_accuracy: 0.9511
Epoch 4/10
48000/48000 [==============================] - 4s 90us/step - loss: 0.1702 - accuracy: 0.9509 - val_loss: 0.1705 - val_accuracy: 0.9528
Epoch 5/10
48000/48000 [==============================] - 4s 93us/step - loss: 0.1661 - accuracy: 0.9525 - val_loss: 0.1678 - val_accuracy: 0.9531
Epoch 6/10
48000/48000 [==============================] - 4s 92us/step - loss: 0.1622 - accuracy: 0.9532 - val_loss: 0.1637 - val_accuracy: 0.9542
Epoch 7/10
48000/48000 [==============================] - 4s 86us/st

In [3]:
# Experiment 2: Changing Batch Size
BATCH_SIZE = 64  # Reduce batch size

model.compile(loss='categorical_crossentropy',
              optimizer=OPTIMIZER,
              metrics=['accuracy'])

history = model.fit(X_train, Y_train,
                    batch_size=BATCH_SIZE, epochs=NB_EPOCH,
                    verbose=VERBOSE, validation_split=VALIDATION_SPLIT)

score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score with batch size 64:", score[0])
print('Test accuracy with batch size 64:', score[1])


Train on 48000 samples, validate on 12000 samples
Epoch 1/10
48000/48000 [==============================] - 7s 142us/step - loss: 0.1446 - accuracy: 0.9581 - val_loss: 0.1502 - val_accuracy: 0.9578
Epoch 2/10
48000/48000 [==============================] - 8s 173us/step - loss: 0.1388 - accuracy: 0.9601 - val_loss: 0.1440 - val_accuracy: 0.9593
Epoch 3/10
48000/48000 [==============================] - 10s 219us/step - loss: 0.1330 - accuracy: 0.9625 - val_loss: 0.1416 - val_accuracy: 0.9598
Epoch 4/10
48000/48000 [==============================] - 26s 533us/step - loss: 0.1275 - accuracy: 0.9638 - val_loss: 0.1371 - val_accuracy: 0.9619
Epoch 5/10
48000/48000 [==============================] - 14s 301us/step - loss: 0.1226 - accuracy: 0.9654 - val_loss: 0.1333 - val_accuracy: 0.9619
Epoch 6/10
48000/48000 [==============================] - 16s 341us/step - loss: 0.1181 - accuracy: 0.9669 - val_loss: 0.1304 - val_accuracy: 0.9620
Epoch 7/10
48000/48000 [==============================] - 

# Analysis of Accuracy Rate Changes

## Initial Setup
- **Test Accuracy**: 94.63%%

## Experiment 1: Varying Number of Epochs
- **Test Accuracy**: 95.36%
- **Changes**: Reducing the number of epochs from 20 to 10 resulted in a slight decrease in test accuracy. Fewer epochs mean less training time, potentially leading to underfitting as the model may not have fully learned from the data.

## Experiment 2: Changing Batch Size
- **Test Accuracy**: 96.54%
- **Changes**: Reducing the batch size from 128 to 64 resulted in a noticeable change in test accuracy. Smaller batch sizes often lead to more frequent updates to the model weights, which can result in faster convergence but also increased noise, impacting overall accuracy.

### Conclusion
- Adjusting these parameters showcases their impact on the model's performance. Finding optimal values requires balancing between training time, model convergence, and accuracy on validation and test sets.
